README.md
# FSX web ATC #

Interactive web map for FSX. Shows navigation aids, air and ground traffic.
Data is taken from [fsx-saas](https://marcosox.github.io/fsx-saas), an HTTP server that exposes FSX simconnect data.

![alt text][screenshot-main]

[screenshot-main]: screenshots/5-air.png "Typical setup for a radar ATC position"

## Installation  

In order to use fsx-web-atc you will need [fsx-saas](https://marcosox.github.io/fsx-saas/) connected to a running FSX 
(check the simconnect notes of fsx-saas for more informations on how to connect to remote FSX instances)

## fsx-web-atc installation
To install fsx-web-atc, download the repository tree into a folder and
serve it from your favourite webserver (e.g. apache).  
No configuration is required if fsx-saas is running on the same host,
 otherwise you can change its address in the config dialog. 

## Usage
The main interface of fsx-web-atc is a world map that shows air, ground and water traffic.

![alt text][screenshot-map]

[screenshot-map]: screenshots/1-main.png "main map"

In the upper right corner, there's the layer selector, where the base map layer and the included overlays can be chosen.

![alt text][screenshot-switcher]

[screenshot-switcher]: screenshots/2-switcher.png "The layer switcher can be used to change what's shown on the map"

In the lower left corner, there's the option panel.

![alt text][screenshot-config]

[screenshot-config]: screenshots/4-config.png "The config dialog"

By changing the layers the view can be adapted for a ground view or an approach/departure position:

![alt text][screenshot-ground]

[screenshot-ground]: screenshots/3-ground.png "A ground position setup"

### Additional layers
Additional overlays can be made available in kml or image format. To add an overlay:
#### KML
1. Place your kml file in a new folder under `charts/kml/`
2. Extend the `KMLOverlayLayers` array in `config.js` by adding a new block:

        var KMLOverlayLayers = [
            {
                name: "LIRF distance rings (kml) - 5 and 10 nm",
                url: "LIRF/Distance_rings.kml"
            }, {
                name: "My new layer: KJFK chart",   // <---- name shown in layer switcher
                url: "KJFK/Ground_chart.kml"    // <---- path to kml file relative to /charts/kml
            }
        ];
        
#### Static images
1. Place your image (any browser supported format should go) in a new folder under `charts/images/`
2. Extend the `imageOverlayLayers` array in `config.js` by adding a new block:

        var imageOverlayLayers = [
            {
                    name: "LIRF AD chart",
                    attribution: "AIP 2017",
                    url: "LIRF/LIRF.png",
                    projection: "UTM",
                    imageExtent: [1359250, 5127500, 1368650, 5139180]
                }, {
                    name: "KJFK AD chart",  // <---- name shown in layer switcher
                    attribution: "marcosox",    // <---- attribution text
                    url: "KJFK/KJFK Ground chart.png",   // <---- url, relative to img/charts/
                    projection: "UTM",  // <---- projection, UTM is meters
                    imageExtent: [1346260, 5101530, 1355720, 5108987]    // <---- this is with correct width:height ratio, best alignment on upper right corner
                }
        ];

## Feedback and contacts
If you think there is a bug, or something is missing or wrong with the documentation/support files, feel free to [open an issue].

## License
This software is released under the LGPL V3 license.
A copy is included in the LICENSE file


[open an issue]: https://github.com/marcosox/fsx-web-atc/issues
[releases page]: https://github.com/marcosox/fsx-web-atc/releases

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- Cose aggiunte
### Changed
- Cose cambiate in funzionalita'
### Deprecated
- Dichiarazioni di funzionalita' deprecate, che verranno rimosse in seguito
### Removed
- Deprecati che vengono rimossi
### Fixed
- Bug risolti
### Security
- Note di sicurezza

## [1.0.0] - 2016-10-19
...ripeti...

## [0.2.0] - 2015-12-03
...ripeti...

## [0.1.0] - 2014-11-26
...ripeti...


[Unreleased]: https://github.com/marcosox/project/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/marcosox/project/compare/v0.3.0...v1.0.0
[0.3.0]: https://github.com/marcosox/project/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/marcosox/project/compare/v0.1.0...v0.2.0

js/map.js
// global constants
// feature id prefixes
var aircraft_id_prefix = "aircraft_";
var boat_id_prefix = "boat_";
var vehicle_id_prefix = "vehicle_";
var track_id_prefix = "track_";
// feature layers list
// order here defines layers z-index. Place them from lowest to highest
var featureLayers = {
    tracks: {
        name: "Aircraft tracks"
    }, airports: {
        name: "airports"
    }, NDB: {
        name: "NDB"
    }, VOR: {
        name: "VOR"
    }, waypoints: {
        name: "waypoints"
    }, boats: {
        name: "boats"
    }, vehicles: {
        name: "vehicles"
    }, groundAircrafts: {
        name: "Aircrafts (parked)"
    }, taxiAircrafts: {
        name: "Aircrafts (taxi)"
    }, aircrafts: {
        name: "Aircrafts (air)"
    }
};

// global variables
var map;
var mousePositionControl;
var refreshLoopID;
var errorCount = 0;
var errorCountThreshold = 10;

/**
 * Custom ol.Map control class for the config button
 * @type {{}}
 */
window.app = {};
app.OpenConfigControl = function (opt_options) {

    var options = opt_options || {};

    var button = document.createElement('button');
    button.innerHTML = "settings";

    var this_ = this;

    function openConfigDialog() {
        $("#configDialog").dialog("open");
    }

    button.addEventListener('click', openConfigDialog, false);
    button.addEventListener('touchstart', openConfigDialog, false);

    var element = document.createElement('div');
    element.className = 'open-config-button ol-unselectable ol-control';
    element.appendChild(button);

    ol.control.Control.call(this, {
        element: element,
        target: options.target
    });
};

/**
 * Persists the app config in local storage
 */
function saveConfig() {
    localStorage.setItem("fsx-web-atc.config", JSON.stringify(config));
    localStorage.setItem("fsx-web-atc.theme", JSON.stringify(styles));
}

/**
 * Retrieves the app config from local storage
 */
function loadConfig() {
    if (localStorage.getItem("fsx-web-atc.config") !== null) {
        config = JSON.parse(localStorage.getItem("fsx-web-atc.config"));
    }
    if (localStorage.getItem("fsx-web-atc.theme") !== null) {
        styles = JSON.parse(localStorage.getItem("fsx-web-atc.theme"));
    }
}

/**
 * Setups a spinner and links it to a config variable
 * @param id html element id
 * @param defaultValue default initial value
 * @param callback callback to run when value changes. Accepts one argument, the widget value.
 */
function setupSpinner(id, defaultValue, callback) {

    var spinner = $(id).spinner({
        stop: function (evt) {
            callback(parseInt(evt.target.value));
            saveConfig();
        }
    });
    spinner[0].value = defaultValue;
}

/**
 * Setups a checkbox and links it to a config variable
 * @param id html element id
 * @param defaultValue default initial value
 * @param callback callback to run when value changes. Accepts one argument, the widget value.
 */
function setupCheckbox(id, defaultValue, callback) {
    var checkbox = $(id);
    checkbox[0].checked = defaultValue;
    checkbox.click(function (evt) {
        callback(evt.target.checked);
        saveConfig();
    });
}

/**
 * Setups a text field and links it to a config variable
 * @param id html element id
 * @param defaultValue default initial value
 * @param callback callback to run when value changes. Accepts one argument, the widget value.
 */
function setupTextField(id, defaultValue, callback) {
    var textField = $(id);
    textField[0].value = defaultValue;
    textField.click(function (evt) {
        callback(evt.target.value);
        saveConfig();
    });
}

/**
 * Setups a color box and links it to a config variable
 * @param id html element id
 * @param defaultValue default initial value
 * @param callback callback to run when value changes. Accepts one argument, the widget value.
 */
function setupColorBox(id, defaultValue, callback) {
    var colorBox = $(id);
    colorBox[0].value = defaultValue;
    colorBox.change(function (evt) {
        callback(evt.target.value);
        saveConfig();
    });
}

/**
 * Setups a dropdown and links it to a config variable
 * @param id html element id
 * @param defaultValue default initial value
 * @param callback callback to run when value changes. Accepts one argument, the widget value.
 */
function setupSelectDropdown(id, defaultValue, callback) {
    var dropDown = $(id);
    dropDown[0].options.selectedIndex = dropDown.find("option[value=" + defaultValue + "]")[0].index;
    dropDown[0].onchange = function () {
        callback(this.options[this.options.selectedIndex].value);
        saveConfig();
    };
}

/**
 * Initialize the config dialog controls
 */
function setupConfigControls() {
    loadConfig();
    $('#config-tabs').tabs();
    setupTextField("#backendAddress", config.backend.address, function (val) {
        config.backend.address = val;
    });
    setupSpinner("#flightPathlength", styles.aircrafts.flightPathLength, function (val) {
        styles.aircrafts.flightPathLength = val;
    });
    setupCheckbox("#showBoatLabels", config.showBoatLabels, function (val) {
        config.showBoatLabels = val;
    });
    setupCheckbox("#showVehicleLabels", config.showVehicleLabels, function (val) {
        config.showVehicleLabels = val;
    });
    setupCheckbox("#followUserAircraft", config.followUserAircraft, function (val) {
        config.followUserAircraft = val;
    });
    setupColorBox("#aircraftColor", styles.aircrafts.iconColor, function (val) {
        styles.aircrafts.iconColor = val;
    });
    setupColorBox("#userAircraftColor", styles.aircrafts.userIconColor, function (val) {
        styles.aircrafts.userIconColor = val;
    });
    setupColorBox("#userAircraftLabelColor", styles.aircrafts.userLabelColor, function (val) {
        styles.aircrafts.userLabelColor = val;
    });
    setupColorBox("#aircraftLabelColor", styles.aircrafts.labelColor, function (val) {
        styles.aircrafts.labelColor = val;
    });
    setupColorBox("#boatColor", styles.boats.iconColor, function (val) {
        styles.boats.iconColor = val;
    });
    setupColorBox("#boatLabelColor", styles.boats.labelColor, function (val) {
        styles.boats.labelColor = val;
    });
    setupColorBox("#vehicleColor", styles.vehicles.iconColor, function (val) {
        styles.vehicles.iconColor = val;
    });
    setupColorBox("#vehicleLabelColor", styles.vehicles.labelColor, function (val) {
        styles.vehicles.labelColor = val;
    });
    setupColorBox("#flightPathColor", styles.aircrafts.flightPathColor, function (val) {
        styles.aircrafts.flightPathColor = val;
    });
    setupColorBox("#AirportsColor", styles.airports.iconColor, function (val) {
        styles.airports.iconColor = val;
    });
    setupColorBox("#AirportLabelColor", styles.airports.labelColor, function (val) {
        styles.airports.labelColor = val;
    });
    setupColorBox("#VORColor", styles.vor.iconColor, function (val) {
        styles.vor.iconColor = val;
    });
    setupColorBox("#VORLabelColor", styles.vor.labelColor, function (val) {
        styles.vor.labelColor = val;
    });
    setupColorBox("#NDBColor", styles.ndb.iconColor, function (val) {
        styles.ndb.iconColor = val;
    });
    setupColorBox("#NDBLabelColor", styles.ndb.labelColor, function (val) {
        styles.ndb.labelColor = val;
    });
    setupColorBox("#WaypointsColor", styles.waypoints.iconColor, function (val) {
        styles.waypoints.iconColor = val;
    });
    setupColorBox("#WaypointsLabelColor", styles.waypoints.labelColor, function (val) {
        styles.waypoints.labelColor = val;
    });
    setupSelectDropdown("#aircraft-label-type-selector", config.aircraftLabelsType, function (val) {
        config.aircraftLabelsType = val;
    });
    setupSelectDropdown("#ground-speed-display", config.groundSpeedFormat, function (val) {
        config.groundSpeedFormat = val;
    });
    setupSelectDropdown("#air-speed-display", config.airSpeedFormat, function (val) {
        config.airSpeedFormat = val;
    });
}

/**
 * Initializes the map and the rest of the app
 */
function initMap() {

    ol.inherits(app.OpenConfigControl, ol.control.Control);

    var baseTilesGroup = new ol.layer.Group({
        'title': 'Base layer',
        layers: []
    });
    var overlayGroup = new ol.layer.Group({
        'title': 'Overlays',
        layers: []
    });

    map = new ol.Map({
        layers: new ol.Collection([baseTilesGroup, overlayGroup]),
        interactions: ol.interaction.defaults({
            // constrainResolution: true,
            altShiftDragRotate: true,
            dragPan: true,
            rotate: true
        }),
        controls: ol.control.defaults().extend([new app.OpenConfigControl()]),
        target: 'ol-map',
        view: new ol.View({
            center: ol.proj.fromLonLat(config.initialPosition),
            // zoomFactor: 1.5,
            minZoom: 3,
            zoom: config.initialZoom
        })
    });

    // setup handler to save last position in local storage
    map.on("moveend", function savePosition(evt) {
        var coords = ol.proj.toLonLat(evt.map.getView().getCenter());
        var position = {
            lon: coords[0],
            lat: coords[1],
            zoom: evt.map.getView().getZoom()
        };
        localStorage.setItem("fsx-web-atc.lastPosition", JSON.stringify(position))
    });

    // setup tile layers
    baseTilesGroup.getLayers().push(new ol.layer.Tile({
            title: "OpenStreetMap",
            type: 'base',
            maxZoom: 19,
            source: new ol.source.OSM()
        })
    );
    for (var i = 0; i < tilesLayers.length; i++) {
        baseTilesGroup.getLayers().push(
            new ol.layer.Tile({
                type: 'base',
                title: tilesLayers[i].name,
                maxZoom: 21,
                source: new ol.source.OSM({
                    url: tilesLayers[i].url,
                    maxZoom: tilesLayers[i].maxZoom,
                    wrapX: true
                })
            })
        );
    }

    // setup overlay ground image layers

    // static images
    for (i = 0; i < imageOverlayLayers.length; i++) {
        var overlay = imageOverlayLayers[i];
        overlayGroup.getLayers().push(
            new ol.layer.Image({
                title: overlay.name,
                source: new ol.source.ImageStatic({
                    attributions: overlay.attribution,
                    url: 'charts/images/' + overlay.url,
                    projection: overlay.projection,
                    imageExtent: overlay.imageExtent
                })
            })
        )
    }

    // KML files
    for (i = 0; i < KMLOverlayLayers.length; i++) {
        var kmlFile = KMLOverlayLayers[i];
        overlayGroup.getLayers().push(
            new ol.layer.Vector({
                title: kmlFile.name,
                source: new ol.source.Vector({
                    url: 'charts/kml/' + kmlFile.url,
                    format: new ol.format.KML()
                })
            })
        )
    }

    // load feature layers from featureLayers
    var layersList = Object.keys(featureLayers);
    for (var k = 0; k < layersList.length; k++) {
        var key = layersList[k];
        featureLayers[key] = new ol.layer.Vector({
            title: featureLayers[key].name,
            source: new ol.source.Vector()
        });
        overlayGroup.getLayers().push(featureLayers[key]);
    }

    // setup map controls
    map.addControl(new ol.control.ScaleLine({
        units: 'nautical'
    }));
    map.addControl(new ol.control.ZoomToExtent({
        tipLabel: "Reset view",
        label: "R",
        extent: map.getView().calculateExtent()
    }));
    mousePositionControl = new ol.control.MousePosition({
        projection: new ol.proj.get("EPSG:4326"),
        coordinateFormat: coordinatesToDMSString
    });
    map.addControl(mousePositionControl);
    map.addControl(new ol.control.ZoomSlider());
    map.addControl(new ol.control.LayerSwitcher({
        target: 'ol-map',
        tipLabel: "Change base layer"
    }));

    // load configuration dialog
    $("#configDialog").load(
        "config.html",
        {},
        function () {
            setupConfigControls();
        }
    ).dialog({
        minWidth: 530,
        autoOpen: false
    });

    // load last position if found in local storage
    if (localStorage.getItem("fsx-web-atc.lastPosition") !== null) {
        var lastPosition = JSON.parse(localStorage.getItem("fsx-web-atc.lastPosition"));
        map.getView().setCenter(ol.proj.fromLonLat([lastPosition.lon, lastPosition.lat]));
        map.getView().setZoom(lastPosition.zoom);
    }

    // start the app
    startLoop();
}

/**
 * Logs an error in the lower status bar
 * @param jqXHR ajax request object
 * @param textStatus status text from ajax response callback
 * @param errorThrown error object thrown by the request
 */
function ajaxErrorFunction(jqXHR, textStatus, errorThrown) {
    console.error(jqXHR, textStatus, errorThrown);

    errorCount++;
    if (errorCount >= errorCountThreshold) {
        stopLoop();
        errorCount = 0;
    }
    switch (textStatus) {
        case "error" : {
            $("#status").text("Error connecting to FSX-saas server - status: " + textStatus + " - response: " + (jqXHR.responseText !== undefined ? jqXHR.responseText : ""));
            break;
        }
        case("timeout"): {
            $("#status").text("Can't connect to FSX-saas server - connection timeout");
            break;
        }
    }
}

/**
 * Returns the specified coordinates as a degrees, minutes, seconds string.
 * Note that this returns coordinates as lat, lon unlike the others
 * @param {number[]} coord coordinates (decimal [lon, lat])
 * @returns {string} output string
 */
function coordinatesToDMSString(coord) {
    var lonD = 0 | coord[0];
    var lonM = Math.round(Math.abs(coord[0]) % 1 * 60);
    var lonS = /*Math.round*/(Math.abs(coord[0]) * 60 % 1 * 60);
    var latD = 0 | coord[1];
    var latM = Math.round(Math.abs(coord[1]) % 1 * 60);
    var latS = /*Math.round*/(Math.abs(coord[1]) * 60 % 1 * 60);
    return (coord[1] < 0 && coord[1] > -1 ? "-" : "")
        + latD + "° "
        + (latM < 10 ? "0" : "") + latM + "' "	// zero padding
        + (latS < 10 ? "0" : "") + latS.toFixed(2) + "\", "	// zero padding
        + (coord[0] < 0 && coord[0] > -1 ? "-" : " ")
        + lonD + "° "
        + (lonM < 10 ? "0" : "") + lonM + "' "	// zero padding
        + (lonS < 10 ? "0" : "") + lonS.toFixed(2) + "\"";	// zero padding
}

/**
 * Stop the requests loop
 */
function stopLoop() {
    window.clearInterval(refreshLoopID);
    $("#status").text("Requests to fsx-saas server stopped.")
}

/**
 * Start the requests loop
 */
function startLoop() {
    // one shot requests
    requestAirports();
    requestVORs();
    requestNDBs();
    requestWaypoints();
    // periodic requests
    refreshLoopID = window.setInterval(requestLoop, config.refreshRate);
    $("#status").text("Requests to fsx-saas server started.")
}

/**
 * Main request loop
 */
function requestLoop() {
    requestAircrafts();
    requestBoats();
    requestVehicles();
}

/**
 * Manually fire all the requests
 */
function refreshView() {
    requestAirports();
    requestVORs();
    requestNDBs();
    requestWaypoints();
    requestAircrafts();
    requestBoats();
    requestVehicles();
}

/**
 * Taken from: https://stackoverflow.com/questions/31297721/how-to-get-a-layer-from-a-feature-in-openlayers-3
 * This is a workaround.
 * Returns the associated layer.
 * @param {ol.Map} map.
 * @return {ol.layer.Vector} Layer.
 */
ol.Feature.prototype.getLayer = function (map) {
    var this_ = this, layer_ = undefined, layersToLookFor = [];
    /**
     * Populates array layersToLookFor with only
     * layers that have features
     */
    var check = function (layer) {
        var source = layer.getSource();
        if (source instanceof ol.source.Vector) {
            var features = source.getFeatures();
            if (features.length > 0) {
                layersToLookFor.push({
                    layer: layer,
                    features: features
                });
            }
        }
    };
    //loop through map layers
    map.getLayers().forEach(function (layer) {
        if (layer instanceof ol.layer.Group) {
            layer.getLayers().forEach(check);
        } else {
            check(layer);
        }
    });
    layersToLookFor.forEach(function (obj) {
        var found = obj.features.some(function (feature) {
            return this_ === feature;
        });
        if (found) {
            //this is the layer we want
            layer_ = obj.layer;
        }
    });
    return layer_;
};

js/boats.js
var boatFields = [
    "id",
    "title",
    "latitude",
    "longitude",
    "altitude",
    "altAgl",
    "airspeed",
    "groundSpeed",
    "bank",
    "heading",
    "rudder",
    "throttle"
];

function boatStyle(resolution) {
    var icon = config.icons.boats.generic;
    var scale = 0.1;
    var title = this.get("title");
    if (title.toLowerCase().indexOf("veh_carrier") !== -1 ||
        title.toLowerCase().indexOf("nimitz") !== -1 ||
        title.toLowerCase().indexOf("eisenhower") !== -1 ||
        title.toLowerCase().indexOf("cvn69") !== -1 ||
        title.toLowerCase().indexOf("cvn68") !== -1) {
        icon = config.icons.boats.carrier;
        scale = 0.3;
    }
    return new ol.style.Style({
        image: new ol.style.Icon(/** @type {olx.style.IconOptions} */ ({
            anchor: [0.5, 0.5],
            anchorXUnits: 'fraction',
            anchorYUnits: 'fraction',
            rotateWithView: true,
            scale: scale,
            rotation: (+this.get("heading") * Math.PI / 180),
            src: icon,
            crossOrigin: 'anonymous',
            color: styles.boats.iconColor
        })),
        text: config.showBoatLabels ? new ol.style.Text({
            text: this.get("id").toString(),
            font: 'bold 12px monospace',
            offsetX: 0,
            offsetY: -8,
            fill: new ol.style.Fill({
                color: styles.boats.labelColor
            })
        }) : undefined
    });
}

function requestBoats() {
    $.ajax({
        dataType: "json",
        url: config.backend.address + config.backend.url.boats,
        mimeType: "application/json",
        success: function (data, textStatus, jqXHR) {
            drawBoats(data);
        },
        error: ajaxErrorFunction
    });
}

function drawBoats(boatsList) {
    for (var i = 0; i < boatsList.length; i++) {
        var boat = boatsList[i];
        var id = boat['id'];
        var coordinates = ol.proj.fromLonLat(/** @type {ol.Coordinate}*/[boat['longitude'], boat['latitude']]);

        var feature = featureLayers.boats.getSource().getFeatureById(boat_id_prefix + id);

        if (feature === null) {
            // create feature
            feature = new ol.Feature({
                    geometry: new ol.geom.Point(coordinates)
                }
            );
            feature.setId(boat_id_prefix + id);
            updateFeatureProperties(feature, boat, boatFields);
            feature.setStyle(boatStyle);

            // add feature to correct layer
            featureLayers.boats.getSource().addFeature(feature);

        } else {
            // update feature
            feature.getGeometry().setCoordinates(coordinates);
            updateFeatureProperties(feature, boat, boatFields);
        }
    }

    // delete old boats
    var deleteArray = [];
    var layerFeatures = featureLayers.boats.getSource().getFeatures();
    for (i = 0; i < layerFeatures.length; i++) {    // for each feature in the layer
        feature = layerFeatures[i];
        id = feature.getId();
        var toBeDeleted = true;
        for (var j = 0; j < boatsList.length; j++) {    // search it in the boats list
            if ((boat_id_prefix + boatsList[j]['id']) === id) {
                toBeDeleted = false;
                break;
            }
        }
        if (toBeDeleted === true) {
            deleteArray.push(feature);
        }
    }
    for (i = 0; i < deleteArray.length; i++) {
        feature = deleteArray[i];
        id = feature.getId().substr(boat_id_prefix.length);
        console.log("removing deleted boat '" + feature.getId() + "'");
        featureLayers.boats.getSource().removeFeature(feature);
    }
}

js/vehicles.js
var vehicleFields = [
    "id",
    "title",
    "atcState",
    "latitude",
    "longitude",
    "groundSpeed",
    "heading",
    "rudder",
    "throttle",
    "windSpeed",
    "windDirection",
    "visibility",
    "ambientTemperature",
    "ambientPressure",
    "barometerPressure"
];

function vehicleStyle(resolution) {
    return new ol.style.Style({
        image: new ol.style.Icon(/** @type {olx.style.IconOptions} */ ({
            anchor: [0.5, 0.5],
            anchorXUnits: 'fraction',
            anchorYUnits: 'fraction',
            rotateWithView: true,
            scale: 0.1,
            rotation: (+this.get("heading") * Math.PI / 180),
            src: config.icons.vehicles.generic,
            crossOrigin: 'anonymous',
            color: styles.vehicles.iconColor
        })),
        text: config.showVehicleLabels ? new ol.style.Text({
            text: this.get("id").toString(),
            font: 'bold 12px monospace',
            offsetX: 0,
            offsetY: -8,
            fill: new ol.style.Fill({
                color: styles.vehicles.labelColor
            })
        }) : undefined
    });
}

function requestVehicles() {
    $.ajax({
        dataType: "json",
        url: config.backend.address + config.backend.url.vehicles,
        mimeType: "application/json",
        success: function (data, textStatus, jqXHR) {
            drawVehicles(data);
        },
        error: ajaxErrorFunction
    });
}

function drawVehicles(vehiclesList) {
    for (var i = 0; i < vehiclesList.length; i++) {
        var vehicle = vehiclesList[i];
        var id = vehicle['id'];
        var coordinates = ol.proj.fromLonLat(/** @type {ol.Coordinate}*/[vehicle['longitude'], vehicle['latitude']]);

        var feature = featureLayers.vehicles.getSource().getFeatureById(vehicle_id_prefix + id);

        if (feature === null) {
            // create feature
            feature = new ol.Feature({
                    geometry: new ol.geom.Point(coordinates)
                }
            );
            feature.setId(vehicle_id_prefix + id);
            updateFeatureProperties(feature, vehicle, vehicleFields);
            feature.setStyle(vehicleStyle);

            // add feature to correct layer
            featureLayers.vehicles.getSource().addFeature(feature);

        } else {
            // update feature
            feature.getGeometry().setCoordinates(coordinates);
            updateFeatureProperties(feature, vehicle, vehicleFields);
        }
    }

    // delete old vehicles
    var deleteArray = [];
    var layerFeatures = featureLayers.vehicles.getSource().getFeatures();
    for (i = 0; i < layerFeatures.length; i++) {    // for each feature in the layer
        feature = layerFeatures[i];
        id = feature.getId();
        var toBeDeleted = true;
        for (var j = 0; j < vehiclesList.length; j++) {    // search it in the vehicles list
            if ((vehicle_id_prefix + vehiclesList[j]['id']) === id) {
                toBeDeleted = false;
                break;
            }
        }
        if (toBeDeleted === true) {
            deleteArray.push(feature);
        }
    }
    for (i = 0; i < deleteArray.length; i++) {
        feature = deleteArray[i];
        id = feature.getId().substr(vehicle_id_prefix.length);
        console.log("removing deleted vehicle '" + feature.getId() + "'");
        featureLayers.vehicles.getSource().removeFeature(feature);
    }
}

js/config.js
/**
 * Tile server layers
 */
var tilesLayers = [
    {
        name: "World Imagery",  // name shown in layer switcher
        url: "http://services.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}",
        maxZoom: 21     // maximum zoom level
    }, {
        name: "World navigation chart",
        url: 'http://services.arcgisonline.com/arcgis/rest/services/Specialty/World_Navigation_Charts/MapServer/tile/{z}/{y}/{x}',
        maxZoom: 10
    }, {
        name: "Dark Matter (labels)",
        url: 'http://{a-f}.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}.png',
        maxZoom: 21
    }, {
        name: "Dark Matter (no labels)",
        url: 'http://{a-f}.basemaps.cartocdn.com/dark_nolabels/{z}/{x}/{y}.png',
        maxZoom: 21
    }, {
        name: "World dark",
        url: 'https://{a-d}.tiles.mapbox.com/v3/mapbox.world-dark/{z}/{x}/{y}.png',
        maxZoom: 11
    }, {
        name: "Stamen toner",
        url: 'http://tile.stamen.com/toner/{z}/{x}/{y}.png',
        maxZoom: 21
    }, {
        name: "Stamen terrain",
        url: 'http://tile.stamen.com/terrain/{z}/{x}/{y}.png',
        maxZoom: 21
    }, {
        name: "Control room",
        url: 'https://{a-d}.tiles.mapbox.com/v3/mapbox.control-room/{z}/{x}/{y}.png',
        maxZoom: 8
    }, {
        name: "Light Gray",
        url: "http://services.arcgisonline.com/arcgis/rest/services/Canvas/World_Light_Gray_Base/MapServer/tile/{z}/{y}/{x}",
        maxZoom: 16
    }, {
        name: "Dark gray",
        url: 'http://services.arcgisonline.com/arcgis/rest/services/Canvas/World_Dark_Gray_Base/MapServer/tile/{z}/{y}/{x}',
        maxZoom: 16
    }
];

/**
 * Chart image overlays
 */
var imageOverlayLayers = [
    {
        name: "LIRF AD chart",  // name shown in layer switcher
        attribution: "AIP 2017",    // attribution text
        url: "LIRF/LIRF.png",   // url, relative to img/charts/
        projection: "UTM",  // projection, UTM is meters
        imageExtent: [1359250, 5127500, 1368650, 5139180]   // image bounding box in meters from 0,0
    }
];

/**
 * KML overlays
 */
var KMLOverlayLayers = [
    {
        name: "LIRF distance rings (kml) - 5 and 10 nm",
        url: "LIRF/Distance_rings.kml"
    }
];

/**
 * App configuration
 */
var config = {
    refreshRate: 1000,	// in ms
    initialPosition: [0.0, 0.0],
    initialZoom: 3,
    groundSpeedFormat: "groundspeed",
    airSpeedFormat: "airspeed",
    aircraftLabelsType: "complete",
    followUserAircraft: true,
    showVehicleLabels: false,
    showBoatLabels: false,
    icons: {
        aircrafts: {
            generic: "img/aircrafts/generic.png"
        },
        boats: {
            generic: "img/boats/boat.png",
            carrier: "img/boats/carrier.png"
        },
        vehicles: {
            generic: "img/vehicles/truck.png"
        },
        airports: "img/square.png",
        vor: "img/vor.png",
        ndb: "img/ndb.png",
        waypoints: "img/waypoint.png"
    },
    backend: {
        address: "http://127.0.0.1:8080",
        url: {
            aircrafts: "/aircrafts",
            vehicles: "/vehicles",
            boats: "/boats",
            airports: "/airports",
            vors: "/vors",
            ndbs: "/ndbs",
            waypoints: "/waypoints"
        }
    }
};

var styles = {
    aircrafts: {
        iconColor: "#FFFFFF",
        userIconColor: '#FF7F00',
        labelColor: "#00FF00",
        userLabelColor: "#FF7F00",
        flightPathColor: "#008000",
        flightPathLength: 9999,
        flightPathWidth: 1
    },
    vehicles: {
        iconColor: "#77240c",
        labelColor: "#9aaa9a"
        // flightPathColor: "#00FF00",  //  TODO: add back
        // flightPathLength: 9999,
        // flightPathWidth: 1
    },
    boats: {
        iconColor: "#0000FF",
        labelColor: "#5050FF"
        // flightPathColor: "#00FF00",  // TODO: add back
        // flightPathLength: 9999,
        // flightPathWidth: 1
    },
    airports: {
        iconColor: "#0000FF",
        labelColor: "#3383ff"
    },
    vor: {
        iconColor: "#00FF00",
        labelColor: "#00FFFF"
    },
    ndb: {
        iconColor: "#0000FF",
        labelColor: "#00FFFF"
    },
    waypoints: {
        iconColor: "#0000FF",
        labelColor: "#00FFFF"
    }
};
